My Home: simplify My Home data layer by migrating to React Query

p-jackson opened the pull request:
## Changes proposed in this Pull Request

In the spirit of trialing React Query more broadly (p4TIVU-9GP-p2) this PR removes the redux store for My Home (`/home/{{ site url }}`) and replaces it with a mixture of React Query and the existing user preferences store. My Home doesn't deal with a lot of client state and is basically a UI that reflects the API, so it felt like a good match for React Query.

* Move the open/close state for the Quick Links card to the user preferences store
* Move the layout fetching and skipping from Redux to ReactQuery
* Small tweak: don't require the `?dev=true` query param when using the `?view=VIEW_ID` option to force a particular view

### Move the open/close state for the Quick Links card to the user preferences store

That's the state for this toggle:
<img width="316" alt="toggle" src="https://user-images.githubusercontent.com/1500769/123233892-90d4a900-d52e-11eb-95de-45dc14bd6b08.png">

Previously the toggle state was being persisted locally in indexdb by virtue of being in the home redux store (`calypso/state/home/`). However we already have a place in redux for this sort of user state (`calypso/state/preferences/`).

One advantage is this preference will now be saved on the server rather than not being remembered across devices. Another advantage is preferences can be inspected using the dev tools:
<img width="692" alt="prefs" src="https://user-images.githubusercontent.com/1500769/123234910-7c44e080-d52f-11eb-8fb8-e6faaa83c577.png">

And the final advantage, is moving the preference allows us to delete the redux store for My Home.

This PR doesn't migrate the user's existing preference into the new store i.e. if the user had collapsed Quick Links, this PR will re-open it until they close it again manually. I initially had logic that migrated the preference across, but I didn't think it was worth the extra bytes and complexity for this particular preference. You can check it out here: #53766

### Move the layout fetching and skipping from Redux to ReactQuery

There's an introductory post on the Calypso P2 p4TIVU-9AF-p2, but the short version is that React Query uses hooks to manage server state. It deals with loading state, caching, exponential backoff after errors etc. Basically the stuff that `calypso/data-layer/` does in a Redux way, React Query does in a hooks way.

It's especially suited when a UI doesn't have any _real_ application state of its own, it's just showing the servers state. And I realised that this matched My Home.

The My Home layout data is queried using the `useHomeLayoutQuery()` hook, and views are skipped using the `useSkipCurrentViewMutation()` hook. These hooks are in `calypso/data/home/`, and you can find other React Query hooks in the `data/` folder.

Some implementation notes off the top of my head:
- React Query likes to automatically keep the data up to date, but the home layout endpoint returns a random result each time it's called, so I've used flags to make sure it will only update on user action
- I'm grabbing the `dev` and `view` url params in the hooks so that we don't have to plumb the props all through home components
- New layout requests always override the old layout, so the query key is set up to ensure there's only ever one layout per site

### No longer require the `?dev=true` query param when forcing a specific view

My Home shows the view that the server thinks is most appropriate for the user. For testing purposes it's useful to be able to force a particular view, which can be done with by adding query params to the browser url bar e.g. `/home/{{ site url }}?dev=true&view=VIEW_WEBINARS`

There didn't seem to be any reason why we should only allow the view to be forced when the `dev` flag is also set. So I've removed that requirement.

The dev flag isn't currently used for anything on the back end, but I'm currently thinking we should retain it because it'd be useful if we ever want to trial new views, or some other new behaviour, and allow a12s to test it without a sandbox.

## Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Everything should operate as before, but here are some things to try:

* Refreshing My Home will display a view that the user is eligable to see
* Skipping a view skips the correct view (check the POST body to ensure the correct view is skipped) and results in a new view being shown
* Check in the network tab that there aren't more requests than expected (no double-requests for example)
* Check the `?dev=true` and `?view=VIEW_ID` query params are correctly sent
* `calypso_customer_home_card_impression` tracks event should still be sent
* Test that the Quick Links toggle state is remembered correctly

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

p-jackson committed:
Add custom query hooks for My Home layout

p-jackson committed:
Fetch and display My Home layout with ReactQuery

p-jackson committed:
Remove home redux store

p-jackson committed:
Store quick links expanded state in preferences tree

Moves the flag for whether the quick links are expanded or not out of
the home store and into the global calypso prefs store.

Doesn't attempt to migrate existing user preference. The expanded state
of the quick links accordian isn't worth the extra bytes and complexity
of the migration code.

jsnajdr commented on client/data/home/use-home-layout-query.ts:
```diff
@@ -0,0 +1,36 @@
+
+const useHomeLayoutQuery = (
+	siteId: number | null,
+	isDev?: boolean,
```
Looking at the backend code, the `isDev` parameter is not currently used anywhere. Maybe we could remove it? It seems that it was added in #42695 only to enable the "forced view" mode. But we could simply enable that for all environments, and remove even the `home/layout-dev` flag. If someone adds manually a `/home?view=abc` param to their URL in production, it doesn't hurt if the desired view is really used, does it? It won't happen by accident.

p-jackson replied:
You're right, I've just noticed it's not used anywhere. I thought about removing it, but it also seemed like it could come in useful later. I think it's there so we can try out new views&mdash;or new logic for which view to display&mdash;and allow a12s to try it out without a sandbox. That feels like it could still be useful to the pod.

> If someone adds manually a `/home?view=abc` param to their URL in production, it doesn't hurt if the desired view is really used, does it? It won't happen by accident.

I agree with this though, I don't think the `dev` param should be required for the `view` param to work. So I've changed that.

jsnajdr commented on client/data/home/use-home-layout-query.ts:
```diff
@@ -0,0 +1,36 @@
+	return useQuery(
+		// Don't include `isDev` or `forcedView` in query key, we want every
+		// caller to have the same idea of what the current view is.
+		[ 'home-layout', siteId ],
```
The cache key should certainly include the `forcedView` param. That params changes the meaning of the response. It's either:
- show me the server-recommended view for this site (determined by your plan, site age, checklist progress...)
- show me one specific view for the site

p-jackson replied:
I tried including `isDev` and `forcedView` in the key, but it didn't end up working well. I always want new layouts to replace old layouts, regardless of whether it was originally requested with a forced view or not.

For example, I use `?view=VIEW_PODCASTING` to load a specific view, and then skip that view. When calling the skip endpoint it returns the view to show now, so I replace the old state with `queryClient.setQueryState()`. But in order to overwrite the old state I need to remember the old view ID so that I overwrite the correct cache key. What if I skip again? Would I have to remember the old-old view ID so I can update the cache state?

It also made it difficult to query what the "current view" was. In order to call `queryClient.getQueryState(['home-layout', siteId, ...])` I need to know all parts of the cache key. But now it's chicken and egg: I need the current view ID so that I can look up the layout data and retrieve the current view ID.

> That params changes the meaning of the response

I think that's true from the API's point of view, but from the client's point of view they both mean the same thing: "the one view to show to the user"

jsnajdr replied:
OK that makes sense. There is only one `[ 'home-layout', siteId ]` response per site and `forcedView` and `isDev` are just debug flags that don't justify a separate key.

jsnajdr commented on client/data/home/use-skip-home-view-mutation.ts:
```diff
@@ -0,0 +1,57 @@
+			} ),
+		{
+			onSuccess() {
+				queryClient.invalidateQueries( [ 'home-layout', siteId ] );
```
Invalidating the query means that React Query will immediately refetch the view, with a second `/home/layout` REST request. But we just got the updated view in the response. All we need to do is to update the data for the query key using the response. We don't need a second request. You want to do something like:
```js
onSuccess( data ) {
  queryClient.setQueryData( [ 'home-layout', siteId ], data );
}
```
Something similar was recently done in the `useVatDetails` hook.

p-jackson replied:
Ah great! I was copy/pasting from `useRemoveFollowerMutation()` without fully understanding the difference so thanks.

jsnajdr commented on client/my-sites/customer-home/cards/actions/quick-links/index.jsx:
```diff
@@ -329,8 +329,8 @@ const mapDispatchToProps = {
-	expand: expandHomeQuickLinks,
-	collapse: collapseHomeQuickLinks,
+	expand: savePreference.bind( null, 'homeQuickLinksToggleStatus', 'expanded' ),
+	collapse: savePreference.bind( null, 'homeQuickLinksToggleStatus', 'collapsed' ),
```
Stylistically, writing this as `() => savePreference( 'homeQuickLinksToggleStatus', 'expanded' )` makes it more clear what the `expand` parameters are and what it does.

jsnajdr commented on client/my-sites/customer-home/cards/actions/quick-links/index.jsx:
```diff
@@ -25,7 +25,7 @@ import { getDomainsBySiteId } from 'calypso/state/sites/domains/selectors';
 import { bumpStat, composeAnalytics, recordTracksEvent } from 'calypso/state/analytics/actions';
 import ActionBox from './action-box';
 import isHomeQuickLinksExpanded from 'calypso/state/selectors/is-home-quick-links-expanded';
```
Now when we toggle the Quick Links by direct calls to `savePreference`, this selector could be also inlined as a direct `getPreference` selector call.

jsnajdr commented on client/my-sites/customer-home/cards/notices/celebrate-notice/index.jsx:
```diff
@@ -47,7 +48,7 @@ const CelebrateNotice = ( {
 	const showNextTask = () => {
 		setIsLoading( true );
-		dispatch( skipViewHomeLayout( siteId, currentView ) );
+		skipHomeView( siteId, currentView );
```
`useCurrentView` is only ever used to get the `skipHomeView`'s second parameter. Then why not integrate it directly into that hook?

Also, the current code is fragile. What if `showNextTask` is called before the `useHomeLayoutQuery` finished fetching? Then `currentView` will be `undefined`.

p-jackson replied:
I've moved the lookup for the current view so it happens within the `useSkipCurrentViewMutation()` hook.

> Also, the current code is fragile. What if `showNextTask` is called before the `useHomeLayoutQuery` finished fetching? Then `currentView` will be `undefined`.

My Home doesn't render _anything_ until `useHomeLayoutQuery` has finished, which means this button (and the ability to call `showNextTask`) won't appear until the current view is a valid value. But to be defensive I've protected the skip call like this:
```typescript
if ( siteId && viewName ) {
	mutate( { siteId, viewName, reminder } );
}
```

Does that address the fragility?

jsnajdr commented on client/my-sites/customer-home/locations/secondary/learn-grow/index.jsx:
```diff
@@ -2,7 +2,7 @@
  */
 import React, { useEffect } from 'react';
-import { connect } from 'react-redux';
+import { connect, useSelector } from 'react-redux';
```
As a rule of thumb, it's a bit suspicious if one module uses both `connect` and the hook API. The component could be easily rewritten to avoid `connect` altogether:
```js
const dispatch = useDispatch();
const siteId = useSelector( getSelectedSiteId );
const cards = useLearnGrowCards( siteId );

useEffect( () => {
  if ( cards ) {
    dispatch( trackCardImpressions( cards ) );
  }
}, [ cards, dispatch ] );
```

jsnajdr commented on client/my-sites/customer-home/locations/tertiary/manage-site/index.jsx:
```diff
@@ -2,7 +2,7 @@
  */
 import React, { useEffect } from 'react';
-import { connect } from 'react-redux';
+import { connect, useSelector } from 'react-redux';
```
Same here, `connect` is easy to remove completely.

p-jackson committed:
Remove connect API

There is still uses of both the connect and hook API from react-redux on
customer home. But this removes the connect() API from the most simple
uses that have already been changed as part of this simplification.

jsnajdr commented on client/data/home/home-layout-flags.ts:
```diff
@@ -0,0 +1,17 @@
+}
+
+export function getHomeLayoutFlags(): HomeLayoutFlags {
+	const pageParams = new URLSearchParams( document.location.search.substring( 1 ) );
```
We have abstractions for routing so that we don't need to read query params directly from `document.location`.

There is the `context.query` object that the original code used to figure out the `isDev` and `forcedView` props.

And there's also the `getCurrentQueryArguments` Redux selector -- the current route information is also in Redux.

p-jackson replied:
I like your example below that uses the `getCurrentQueryArguments` selector 👍

jsnajdr commented on client/data/home/use-home-layout-query.ts:
```diff
@@ -0,0 +1,46 @@
+				{
+					...( isDev && { dev: true } ),
+					...( forcedView && { view: forcedView } ),
+				}
```
If `getHomeLayoutFlags` returns `{ dev: undefined }`, the `dev` key will be omitted from the actual query string. That way, you could directly pass the return value without structuring, destructuring, and structuring objects again.

jsnajdr commented on client/data/home/use-home-layout-query.ts:
```diff
@@ -0,0 +1,46 @@
+			// The `/layout` endpoint can return a random view. Disable implicit refetches
+			// so the view doesn't change without some user action.
+			staleTime: Infinity,
+			refetchOnMount: fetchOnMount && 'always',
```
The original `<QueryHomeLayout/>` always fetches on mount and only on mount. To reproduce that, `refetchOnMount` should be always `always`. You might need to set `refetchInterval: false`, too.

p-jackson replied:
> The original `<QueryHomeLayout/>` always fetches on mount and only on mount.

That's true, but we also only included `QueryHomeLayout` in one place. The `useHomeLayoutQuery` hook is included in 3 places and I only want the request to happen once (always using `always` results in two requests, I think the 2nd two mounts are merged into one request because they mount in the same render pass).

I'm replicating the old behaviour by having the component that used to include `<QueryHomeLayout/>` now explicitly saying `fetchOnMount=true`, and the other places that used to just select the layout data leave `fetchOnMount` as `false`.

> You might need to set `refetchInterval: false`, too.

`false` is the default, but it wouldn't hurt to be explicit in case the default ever changes. I am trying to say "don't leave it up to React Query to do what it thinks is best" after all.

jsnajdr replied:
> The `useHomeLayoutQuery` hook is included in 3 places and I only want the request to happen once

I see, the other two places are the ones that used to just passively call the selector. The best way to make a passive React Query like this is probably:
```js
useHomeLayoutQuery( siteId, { enabled: false } );
```
This hook will only read the current cached state and will not actively trigger anything.

I think it's better because it disables all fetches and retries with just one flag, instead of carefully disabling all methods (on mount, on window focus, on interval, ...) one by one.

p-jackson replied:
I've switched the `useHomeLayoutQuery()` hook over from not querying by default, to querying by default. It feels like our custom queries should have some sort of convention when a situation like this arises, and I like the idea of forwarding an `enable` option to `useQuery()`. Using the same flags as ReactQuery is a good convention.

jsnajdr commented on client/data/home/use-skip-current-view-mutation.ts:
```diff
@@ -0,0 +1,66 @@
+
+	const skipCurrentView = useCallback(
+		( siteId: number, reminder: ReminderDuration ) => {
+			const data = queryClient.getQueryData( [ 'home-layout', siteId ] );
```
This code relies on someone else to fire the `useHomeLayoutQuery` query and save the response to cache. If nobody fires the query or if `skipCurrentView` is called before the query response arrives, the mutation will silently do nothing.

Code like this would be more reliable:
```js
const layout = await wpcom.req.get( '/sites/${ id }/home/layout' );
await wpcom.req.post( '/sites/${ id }/home/layout/skip', { view: layout.view_name } );
```
Only we don't want to use `wpcom.req` directly, but wrapped in React Query so that cached data are used if available.

This code makes sure that the `layout` is really fetched, waits if it needs to wait, and always performs the mutation with the right `view` parameter.

jsnajdr replied:
I was able to make this mutation more reliable by fetching the current view in the mutation function:
```js
const mutation = useMutation(
  async ( { reminder } ) => {
    const data = await queryClient.fetchQuery(
      [ 'home-layout', siteId ],
      () => fetchHomeLayout( siteId, query ),
      { staleTime: Infinity }
    );
    const view = data.view_name;
    return await wp.req.post( `/sites/${ siteId }/skip`, { view, reminder } );
  },
  {
    onSuccess( data ) {
      queryClient.setQueryData( ... );
    }
  }
);
```

The `fetchHomeLayout` function is extracted and exported from `useHomeLayoutQuery`, being reused in both hooks.

The `queryClient.fetchQuery` will return data from cache, if it was already fetched and stored, or will issue the fetch request itself.

The `staleTime: Infinity` option makes sure that `fetchQuery` never re-fetches if the data are already in cache. The default `staleTime` is 0, causing a refetch on every call.

p-jackson replied:
Nice! I was thinking of calling `useHomeLayoutQuery()` from within the body of `useSkipCurrentViewMutation()`, but I didn't want to pre-emptively fetch until someone had actually called the mutate function. This gets around this nicely. I very much like that ReactQuery has a non-React API like this.

jsnajdr commented on client/my-sites/customer-home/locations/tertiary/manage-site/index.jsx:
```diff
@@ -55,16 +58,14 @@ const ManageSite = ( { cards, trackCards } ) => {
-	const layout = getHomeLayout( state, siteId );
+function useManageSiteCards() {
+	const siteId = useSelector( getSelectedSiteId );
+	const { data: layout } = useHomeLayoutQuery( siteId );
```
There could be a `useHomeLayoutFlags` hooks like:
```js
function useHomeLayoutFlags() {
  const query = useSelector( getCurrentQueryArguments );
  return {
    dev: config.isEnabled( 'home/layout-dev' ) ? query.dev : undefined,
    view: query.view,
  };
}

jsnajdr reviewed (commented):
I experimented with removing the race condition from the `useSkipCurrentViewMutation` and found a solution that works very nicely.

After fixing this, the PR is ready to merge from my point of view :+1:

tyxla commented on client/data/home/use-home-layout-query-params.ts:
```diff
@@ -0,0 +1,24 @@
+	const { dev, view } = useSelector( getCurrentQueryArguments ) as any;
+
+	return {
+		dev: config.isEnabled( 'home/layout-dev' ) || dev === 'true' || undefined,
```
I know this was like that before, but I'd like to point out that it can be confusing to have it enabled by either a feature flag or a query argument. Usually, when I enable a feature flag, I expect that the related functionality will be triggered. Likewise, if I disable it, I expect the related functionality will not be triggered - but in this instance, the opposite will happen. Maybe that's just fine, but I just wanted to point out it feels odd from an external perspective 😉

p-jackson replied:
Now that you mention it I would expect the explicit query param to take precedence over the config i.e. I can use `?dev=false` to override the config. The short answer is that this situation hasn't come up.

[We thought about removing the dev flag](https://github.com/Automattic/wp-calypso/pull/53207#discussion_r654198636). I want to keep it around because I can see it being useful for future backend changes, but given it's not used for anything ATM we can change the logic to whatever makes sense.

I'm going to make the query param take precedence.

tyxla commented on client/data/home/use-home-layout-query.ts:
```diff
@@ -0,0 +1,46 @@
+
+			// The `/layout` endpoint can return a random view. Disable implicit refetches
+			// so the view doesn't change without some user action.
+			staleTime: Infinity,
```
Are there any conditions for which we'd want to re-enable re-fetching? For example, if another user has logged in (and uses the same site), or if the locale has changed, or something else, non-site related? Perhaps it's just fine if we keep the same view for the next user who logged in?

jsnajdr replied:
The query cache gets destroyed on page reload, and logging in as another user always involves a full reload. Calypso can't switch a user at runtime. Even if the cache starts being persisted to local storage, it will be keyed by user ID.

p-jackson replied:
> Are there any conditions for which we'd want to re-enable re-fetching? For example, if another user has logged in (and uses the same site)

Really interesting question about how resilient we think hooks/components should be. It made me think about what we'd do if we wanted our components to render in storybook, would we have to assume that the Calypso layout wasn't there to do a bunch of the work for us? 🤔 I imagine in storybook we'd clear the cache between navigating to different stories.
I think this is similar to how in most Calypso code it's ok to assume the selected site data is available in Redux, even though to be fully resilient you could argue each component should request it just to be sure.

Just assuming that the query cache will be destroyed during logout feels like the right trade off to me.

jsnajdr replied:
I see the `/home/layout` endpoint being a special case because we want to have a very precise control over when we fetch from the endpoint. Because the endpoint returns randomized results, we want to fetch from it once and only once for each Home pageview. Re-fetching, which is usually a harmless way to get fresh data, is undesired here.

The default React Query behavior is to fetch the data on i) mount ii) window focus iii) reconnect. Then mark it immediately as stale (`staleTime = 0`) and refetch on the next mount/focus/reconnect. But for `/home/layout`, we want to do the fetch only on one specific mount and nowhere else.

A possible improvement of the current situation, to make sure that we fetch only once on one specific mount, would be to call `useHomeLayoutQuery` only at that one place, and drill down the current `layout` down as a prop or in a context.

tyxla reviewed (commented):
This is wonderful work 👍  

I don't have much to add to what Jarda did already, just a couple of comments. 

Great use case of `react-query` btw 👍

p-jackson committed:
The dev query param overrides the config

Using the `?dev=false` query param now overrides the `home/layout-dev`
config flag. This could be useful when developing locally and you want
to quickly test the production (i.e. non-dev) behaviour of the API

jsnajdr commented on client/data/home/use-skip-current-view-mutation.ts:
```diff
@@ -0,0 +1,64 @@
+		async ( { reminder } ) => {
+			const data = await queryClient.fetchQuery(
+				getCacheKey( siteId ),
+				() => fetchHomeLayout( siteId, { dev } ),
```
Here you should pass the entire `query` returned from `useHomeLayoutQueryParams()`, not just the `dev` part. That guarantees that both places where we execute the query (the hook and this async function) use exactly the same Query Function and that it doesn't matter which one is called first.

p-jackson replied:
Thanks for catching, I got confused because we don't want to send the `view` param to the skip API, so didn't think we'd want it here either.

> That guarantees that both places where we execute the query (the hook and this async function) use exactly the same Query Function and that it doesn't matter which one is called first.

👍

jsnajdr reviewed (approved):
Just one nit about the `query` passed to `fetchHomeLayout`. Once it's fixed, this is ready to go :+1:

Thanks for working on this! We've seen and solved many React Query problems in this PR that we didn't encounter before.
